Add JS API to attach metadata to an element during autofill

rniwa opened the pull request:
#### 4a788cf43c3144d67fdfc2b7c2cae28ec74463a2
<pre>
Add JS API to attach metadata to an element during autofill
<a href="https://bugs.webkit.org/show_bug.cgi?id=283247">https://bugs.webkit.org/show_bug.cgi?id=283247</a>

Reviewed by Wenson Hsieh.

This PR adds a new content world configuration to expose JS API on Element to attach user info,
and pass this information to WKUIDelegatePrivate&apos;s _webView:didClickAutoFillButtonWithUserInfo
during autofill.

It also introduces &quot;webkitautofillrequest&quot; event when the user activates an autofill button in
an input element. The event only fires in a content world where autofill capability is enabled.

* Source/WebCore/bindings/js/DOMWrapperWorld.h:
(WebCore::DOMWrapperWorld::setAllowElementUserInfo):
(WebCore::DOMWrapperWorld::allowElementUserInfo const):
* Source/WebCore/dom/Element.cpp:
(WebCore::Element::setUserInfo):
(WebCore::Element::userInfo const):
* Source/WebCore/dom/Element.h:
* Source/WebCore/dom/Element.idl:
* Source/WebCore/dom/ElementRareData.cpp:
* Source/WebCore/dom/ElementRareData.h:
(WebCore::ElementRareData::userInfo const):
(WebCore::ElementRareData::setUserInfo):
(WebCore::ElementRareData::useTypes const):
* Source/WebCore/dom/Event.cpp:
(WebCore::Event::Event):
* Source/WebCore/dom/Event.h:
(WebCore::Event::isAutofillEvent):
(WebCore::Event::setIsAutofillEvent):
* Source/WebCore/dom/EventNames.json:
* Source/WebCore/dom/EventTarget.cpp:
(WebCore::EventTarget::innerInvokeEventListeners):
* Source/WebCore/dom/Node.cpp:
(WebCore::stringForRareDataUseType):
* Source/WebCore/dom/NodeRareData.h:
* Source/WebCore/html/TextFieldInputType.cpp:
(WebCore::TextFieldInputType::autoFillButtonElementWasClicked):
* Source/WebKit/Shared/ContentWorldData.serialization.in:
* Source/WebKit/Shared/ContentWorldShared.h:
* Source/WebKit/UIProcess/API/APIContentWorld.h:
* Source/WebKit/UIProcess/API/Cocoa/WKContentWorld.mm:
(checkContentWorldOptions):
(+[WKContentWorld _worldWithConfiguration:]):
* Source/WebKit/UIProcess/API/Cocoa/WKContentWorldConfiguration.mm:
(-[_WKContentWorldConfiguration copyWithZone:]):
(-[_WKContentWorldConfiguration encodeWithCoder:]):
(-[_WKContentWorldConfiguration initWithCoder:]):
* Source/WebKit/UIProcess/API/Cocoa/_WKContentWorldConfiguration.h:
* Source/WebKit/WebProcess/InjectedBundle/InjectedBundleScriptWorld.cpp:
(WebKit::InjectedBundleScriptWorld::setAllowElementUserInfo):
* Source/WebKit/WebProcess/InjectedBundle/InjectedBundleScriptWorld.h:
* Source/WebKit/WebProcess/UserContent/WebUserContentController.cpp:
(WebKit::WebUserContentController::addContentWorld):
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.cpp:
(WebKit::userDataFromJSONData):
(WebKit::WebChromeClient::handleAutoFillButtonClick):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/UserContentController.mm:
(-[AutoFillDelegateForElementUserInfo _webView:didClickAutoFillButtonWithUserInfo:]):
(TEST(WKUserContentController, AllowElementUserInfo)):

Canonical link: <a href="https://commits.webkit.org/286826@main">https://commits.webkit.org/286826@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4853bac406f4980295024255bc675973b6454b76

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/77188 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/56223 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/30103 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/81750 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/28470 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/79305 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/65371 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/4519 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/60481 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/18530 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/80255 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/50438 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/66252 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/40783 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/47841 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/23750 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/26793 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/68955 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/24085 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/83175 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/4568 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/3080 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/68755 "Found 2 new test failures: imported/w3c/web-platform-tests/css/css-pseudo/marker-animate-002.html imported/w3c/web-platform-tests/css/css-view-transitions/active-view-transition-pseudo-class-match.html (failure)") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/4724 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/66225 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/68011 "Found 1 new API test failure: /TestWebKit:WebKit.Find (failure)") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/11996 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/10088 "Passed tests") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/11954 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/4515 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/7330 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/4534 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/7969 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/6293 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rr-codes commented on Source/WebKit/UIProcess/API/Cocoa/_WKContentWorldOptions.h:
```diff
@@ -0,0 +1,55 @@
+WK_SWIFT_UI_ACTOR
+NS_SWIFT_SENDABLE
+WK_CLASS_AVAILABLE(macos(WK_MAC_TBA), ios(WK_IOS_TBA), visionos(WK_XROS_TBA))
+@interface _WKContentWorldOptions : NSObject<NSCopying, NSSecureCoding>
```
instead of an interface, should/could this be an NS_OPTIONS enum (& OptionSet in Swift)?

whsieh commented on Source/WebCore/html/TextFieldInputType.cpp:
```diff
@@ -820,7 +820,13 @@ void TextFieldInputType::autoFillButtonElementWasClicked()
         return;
 
-    page->chrome().client().handleAutoFillButtonClick(*element());
+    RefPtr element = this->element();
```
Nit - I think you could move this smart pointer a few lines up, and use it when getting the Page (and for the debug assertion).

rniwa replied:
Good idea. Fixed.

whsieh commented on Source/WebCore/dom/ElementRareData.h:
```diff
@@ -95,6 +95,9 @@ class ElementRareData : public NodeRareData {
     void setAttributeMap(std::unique_ptr<NamedNodeMap>&& attributeMap) { m_attributeMap = WTFMove(attributeMap); }
 
+    String userInfo() const { return m_userInfo; }
+    void setUserInfo(String& userInfo) { m_userInfo = userInfo; }
```
Can we use move semantics for this setter?

```suggestion
    void setUserInfo(String&& userInfo) { m_userInfo = WTFMove(userInfo); }
```

rniwa replied:
Sure.

whsieh commented on Source/WebCore/dom/Element.cpp:
```diff
@@ -2184,6 +2185,24 @@ bool Element::isElementsArrayReflectionAttribute(const QualifiedName& name)
+    if (throwScope.exception())
+        return;
+
+    ensureElementRareData().setUserInfo(serializedData);
```
Could we change the setter to use move semantics?

```suggestion
    ensureElementRareData().setUserInfo(WTFMove(serializedData));
```

rniwa replied:
Done.

rniwa committed:
Add JS API to attach metadata to an element during autofill
https://bugs.webkit.org/show_bug.cgi?id=283247

Reviewed by Wenson Hsieh.

This PR adds a new content world configuration to expose JS API on Element to attach user info,
and pass this information to WKUIDelegatePrivate's _webView:didClickAutoFillButtonWithUserInfo
during autofill.

It also introduces "webkitautofillrequest" event when the user activates an autofill button in
an input element. The event only fires in a content world where autofill capability is enabled.

* Source/WebCore/bindings/js/DOMWrapperWorld.h:
(WebCore::DOMWrapperWorld::setAllowElementUserInfo):
(WebCore::DOMWrapperWorld::allowElementUserInfo const):
* Source/WebCore/dom/Element.cpp:
(WebCore::Element::setUserInfo):
(WebCore::Element::userInfo const):
* Source/WebCore/dom/Element.h:
* Source/WebCore/dom/Element.idl:
* Source/WebCore/dom/ElementRareData.cpp:
* Source/WebCore/dom/ElementRareData.h:
(WebCore::ElementRareData::userInfo const):
(WebCore::ElementRareData::setUserInfo):
(WebCore::ElementRareData::useTypes const):
* Source/WebCore/dom/Event.cpp:
(WebCore::Event::Event):
* Source/WebCore/dom/Event.h:
(WebCore::Event::isAutofillEvent):
(WebCore::Event::setIsAutofillEvent):
* Source/WebCore/dom/EventNames.json:
* Source/WebCore/dom/EventTarget.cpp:
(WebCore::EventTarget::innerInvokeEventListeners):
* Source/WebCore/dom/Node.cpp:
(WebCore::stringForRareDataUseType):
* Source/WebCore/dom/NodeRareData.h:
* Source/WebCore/html/TextFieldInputType.cpp:
(WebCore::TextFieldInputType::autoFillButtonElementWasClicked):
* Source/WebKit/Shared/ContentWorldData.serialization.in:
* Source/WebKit/Shared/ContentWorldShared.h:
* Source/WebKit/UIProcess/API/APIContentWorld.h:
* Source/WebKit/UIProcess/API/Cocoa/WKContentWorld.mm:
(checkContentWorldOptions):
(+[WKContentWorld _worldWithConfiguration:]):
* Source/WebKit/UIProcess/API/Cocoa/WKContentWorldConfiguration.mm:
(-[_WKContentWorldConfiguration copyWithZone:]):
(-[_WKContentWorldConfiguration encodeWithCoder:]):
(-[_WKContentWorldConfiguration initWithCoder:]):
* Source/WebKit/UIProcess/API/Cocoa/_WKContentWorldConfiguration.h:
* Source/WebKit/WebProcess/InjectedBundle/InjectedBundleScriptWorld.cpp:
(WebKit::InjectedBundleScriptWorld::setAllowElementUserInfo):
* Source/WebKit/WebProcess/InjectedBundle/InjectedBundleScriptWorld.h:
* Source/WebKit/WebProcess/UserContent/WebUserContentController.cpp:
(WebKit::WebUserContentController::addContentWorld):
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.cpp:
(WebKit::userDataFromJSONData):
(WebKit::WebChromeClient::handleAutoFillButtonClick):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/UserContentController.mm:
(-[AutoFillDelegateForElementUserInfo _webView:didClickAutoFillButtonWithUserInfo:]):
(TEST(WKUserContentController, AllowElementUserInfo)):

Canonical link: https://commits.webkit.org/286826@main
